saleor/checkout/calculations.py

```diff
@@ -496,6 +510,55 @@ def _fetch_checkout_prices_if_expired(
     return checkout_info, lines
 
 
+@allow_writer()
+def recalculate_discounts(
+    checkout_info: "CheckoutInfo",
+    lines_info: Iterable["CheckoutLineInfo"],
+    database_connection_name: str = settings.DATABASE_CONNECTION_DEFAULT_NAME,
+    force_update: bool = False,
+) -> tuple["CheckoutInfo", Iterable["CheckoutLineInfo"]]:
+    """Recalculate checkout discounts.
+
+    Discounts are recalculated only if force_update is True, or if both discount
+    and price expirations have passed.
+    This updates catalogue promotions, vouchers, and order promotion discounts.
+    """
+    checkout = checkout_info.checkout
+
+    # Do not recalculate discounts in case the checkout prices are still valid, either
+    # discounts or tax prices.
+    if not force_update and (
+        checkout.discount_price_expiration > timezone.now()
+        or checkout.price_expiration > timezone.now()
+    ):
+        return checkout_info, lines_info
+
+    lines = cast(list, lines_info)
+    update_undiscounted_prices(checkout_info, lines)
+
+    soonest_promotion_end_date = (
+        create_or_update_discount_objects_from_promotion_for_checkout(
+            checkout_info, lines, database_connection_name
+        )
+    )
+
+    if soonest_promotion_end_date is not None:
+        checkout.discount_price_expiration = min(
+            soonest_promotion_end_date, timezone.now() + settings.CHECKOUT_PRICES_TTL
+        )
+    else:
+        checkout.discount_price_expiration = (
+            timezone.now() + settings.CHECKOUT_PRICES_TTL
+        )
+
+    checkout.save(
+        update_fields=["discount_price_expiration"],
```

Comment:
If you modified the `discount_price_expiration`, then it means, that you could have different discount objects. In that case I would say that we also need to invalidate `price_expiration`